### ISSUE TITLE
ESC hangs on strict or in ensures clause

### ISSUE THREAD

**Issue by Jankoekenpan** [OpenJML/OpenJML#594]
Hi David,

I've got this method:
```
	/*@
   	@   requires array != null;
   	@   assignable \nothing;
   	@   ensures ( 0 <= \result
   	@   		&& \result < array.length
   	@   		&& array[\result] == 5)
   	@   		| \result == array.length ;
   	@   ensures (\forall int j; 0 <= j && j < \result; array[j] != 5);
   	@*/
	public static /*@pure@*/ int findFirstFive(int[] array) {
		int k = 0;
		
		/*@ loop_invariant
   		@   0 <= k && k <= array.length 
   		@   && (\forall int i; 0 <= i && i < k; array[i] != 5);
   		@  decreases array.length - k;
   		@*/
		for (; k < array.length; k++) {
			if (array[k] == 5) {
				break;
			}
		}
		return k;
	}
```

When I run extended static checking OpenJML doesn't terminate, however when the bitwise OR (`|`) is replaced by its lazy (`||`) counterpart, it checks just fine. I'm using z3 v4.3.2 as the prover. I'm unfamiliar with SMT, so please find the logs attached.

[success.log](https://github.com/OpenJML/OpenJML/files/1551940/success.log)
[fail.log](https://github.com/OpenJML/OpenJML/files/1551941/fail.log)

**Comment by davidcok** [comment]
I will investigate, but in general if there are bit-operations, then the solver needs to use bit vectors to represent integers and solving can take very much longer.

- David

**Comment by davidcok** [comment]
Fixed in 0.8.25 - logical bit-or does not need to use bit-vector SMT logic. The 'hang' is really that bit-vector SMT can take very much longer than mathematical SMT.
